Cache addHoverClass as it is quite expensive

Juice10 opened the pull request:
Problem:
`addHoverClass` is one of the most CPU intensive things we do when rebuilding a snapshot. And especially when scrubbing backwards or browsing a recording by clicking around on the timeline `addHoverClass` gets repeatedly called.

Solution:
This PR caches the results of `addHoverClass` so they only have to be calculated once and after that they are fast.

Juice10 committed:
Cache addHoverClass as it is quite expensive

JohnPhamous commented:
Nice! We had perf issues for this on our fork so we just disabled `addHoverClass`.

Yuyz0112 commented:
LGTM.

Shall we introduce some ways to invalidate the cache? Just for a general design, every cache should have a way to invalidate, otherwise, it is possible to result in memory leak.

Juice10 committed:
add ability to create and manage a cache object

Juice10 commented:
@Yuyz0112 Good point, I refactored it a bit so it will accept a cache object and I added the cache creation/reset to rrweb in this PR: https://github.com/rrweb-io/rrweb/pull/643

Juice10 commented:
Moved to https://github.com/rrweb-io/rrweb/pull/643 due to monorepo merger
